include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -340,39 +339,65 @@ class GPUTracerManager {
 }; /* class GPUTracerManager */
 
 // Base class for a GPU profiler
+template <typename TManager>
 class GPUProfilerBase : public EpProfiler {
  protected:
   GPUProfilerBase() = default;
+  virtual ~GPUProfilerBase() {}
+
   void MergeEvents(std::map<uint64_t, Events>& events_to_merge, Events& events) {
     Events merged_events;
 
     auto event_iter = std::make_move_iterator(events.begin());
     auto event_end = std::make_move_iterator(events.end());
     for (auto& map_iter : events_to_merge) {
       auto ts = static_cast<long long>(map_iter.first);
-      while (event_iter != event_end && event_iter->ts < ts) {
+
+      // find the last occurence of a matching timestamp,
+      // if one exists
+      while (event_iter != event_end &&
+             (event_iter->ts < ts ||
+              (event_iter->ts == ts &&
+               (event_iter + 1) != event_end &&
+               (event_iter + 1)->ts == ts))) {
         merged_events.emplace_back(*event_iter);
         ++event_iter;
       }
 
-      // find the last event with the same timestamp.
-      while (event_iter != event_end && event_iter->ts == ts && (event_iter + 1)->ts == ts) {
-        ++event_iter;
-      }
+      uint64_t increment;
+      uint64_t last_ts;
+      bool copy_op_names = false;
+      std::string op_name;
+      std::string parent_name;
 
+      // Tracers may not use Jan 1 1970 as an epoch for timestamps.
+      // So, we need to adjust the timestamp to something sensible.
```

Comment:
I am not sure how this works actually, can you elaborate a bit?

Reply:
Sure. This code is essentially trying to account for the fact that the timestamps reported by tracers (specifically, the ROCm tracer) for the beginning and end of a kernel are not guaranteed to use the "standard" epoch of Jan 1 1970. I believe the ROCm tracer uses system boot time as the epoch for timestamps (but I couldn't find this documented anywhere). Now, when we subtract the profiling start time (which is the number of ticks from Jan 1 1970), we end up with a large negative value, which messes up the visualization as you observed. This code essentially tries to work around this behavior by ignoring the timestamps from tracers, and filling in something that is logical.

Note that this does not affect the durations of events: those are calculated simply as ` duration = end_time - start_time`. It only affects the start timestamp. I'd argue that despite the inaccuracy, the start time is closer to the actual start time than if we simply used the tracer timestamps.